### ISSUE TITLE
Feature Request: Centralize core output strings

### ISSUE THREAD

**Issue by Griatch** [evennia/evennia#663]
#### Description of the suggested feature and how it is supposed to work for the admin/end user:

The suggested feature is to move hard-coded output strings (things like "Portal has disconnected" or "Lockstring error") out from the respective modules and into new modules `strings.py` in each subpackage of the Evennia library. The `strings.py` module would look as follows (example from the `objects` subpackage): 

``` python

models_db_player_help_text = 'a Player connected to this object, if any.'
models_db_sessid_help_text='csv list of session ids of connected Player, if any.'
# ...
models_set_location_runtimeerror='Error: {self}.location = {destination} creates a location loop.'
```

... and so on. So, the string named according to an ordered schema. The proposed schema here is `modulename_<unique_name>` but it could also be formalized to `modulename_classname_funcname_varname` (although this feels very cumbersome to actually use). The string in question should have formatting markers for use with the `format()` method of strings. In `objects.models`, this string would then be used like this: 

``` python
from evennia.objects import strings
# ...
except Exception, e:
       errmsg = strings.models_set_location_runtimerror.format(self=self.key, location=location)
       logger.log_errmsg(errmsg)
       raise
```

Finally, there should be a mechanism for easily overloading named strings from the game-dir. This could be something as simple as a module `custom_core_strings.py` in `server/conf` where users can specify a given string (they are responsible for adding the correct formatting markers) which will then overload the core version at server start. Each subpackage `strings` just does `import *` from this module as their last actiont their ends, which will overload matching strings with custom ones (the details of this is not clear yet).
#### A list of arguments for why you think this new feature should be included in Evennia:
1. The current core strings are today completely hard-coded into the core library of Evennia. Some of these do occationally appear for the end-user too. Currently for a developer to override them requires to either to manually hack the library core or to find the ones who are under i18n and write a custom language file for them. The first is hackish and something we really don't recommend for maintainability reasons. The latter works but for those who don't want to actually change the language but just change some phrasing this is hardly intuitive. 
   1. The ability to overload core strings via a settings file means it's a lot easier for developers to customize the strings and also for core devs to track the available messages (maybe introduce some more string-reuse).
2. Concerning string-reuse there could also be a "central" `strings` module (directly under `evennia` that other `strings` modules could pull from - these would be common phrases and error messages used repeatedly. This would make it easier to customize in one place (if these are imported into each sub-`strings` module, then they could still be overloaded on a per-subpackage form if desired - maybe. 
3. It would allows us to completely centralize internationalization. The `strings` modules could be run through i18n in one go, grabbing all strings at the same time. It would make it easy to know what was currently covered or not. 
4. Converting to using `.format()` string formatting not only makes strings compliant with i18n (some languages just puts things in a different order), it is also a small step towards cleaner py3 compliance. 
#### Extra information, such as requirements or ideas on implementation:

_Note that this feature does NOT cover the changing of strings in user-modifiable modules, like Typeclasses, Commands and Lockfuncs. This is only for core strings in modules that the user is not supposed to otherwise touch. Other modules already have mechanism for how to customize them, and introducing this system there would just make them harder to read._

Even though making the mechanical system behind it (allowing for overloading, batch-applying i18n) is probably not too hard.

The main work here (once the system is in place) is to convert all existing strings to use the centralized `strings` modules with proper formatting - that means that for every string:
1. Check if an existing string variable can be re-used
2. If not, copy the string to the proper `strings` module and convert it to use `{format}` statements and give it a suitable variable name. 
3. Replace the current position in the main module with the variable and use `.format()` as applicable.
4. Test.

This is one of those projects that many people can be involved in, also people with little Python exprience - it's not hard, just a lot of work.

**Comment by luyijun** [comment]
It's wonderful! Thank you for your hark work. Now we can translate these messages easily.

And could logger provides a get_last_error function or something like it? If an error happens, the caller (the code, not the end-user) can use this function to know what has happened.

**Comment by carenajosh** [comment]
Do you really want to call it strings.py? D:

**Comment by Griatch** [comment]
No work have been done on this.  We are open right alternative suggestions?

**Comment by strikaco** [comment]
It's a noble idea but I think the proposal would make more of a mess than it solves. I've seen this work ok on smaller scripts and applications written by a single developer where you have maybe a few dozen output strings total but at Evennia's size, replacing user-facing strings with variables would greatly impact the readability of code itself. For the benefit of a handful of users who might want to rephrase cherry-picked strings, this would impact all developers.

> The latter works but for those who don't want to actually change the language but just change some phrasing this is hardly intuitive.

But...that's the entire point of the i18n module. If you want the existing strings *translated* into a different *language* or form, that's literally what i18n is for. i18n is a well-documented, well-integrated and very round wheel we'd be remiss to try to reinvent. And hamburger-wrapped `_('Password is invalid')` inline strings are much more readable at scale (and a more commonly understood pattern) than references to `specific_failure_in_a_particular_context_1` singletons in a separate module.

Creating an interface to make it easier to change arbitrary strings via i18n and automate exporting of *.po files would be a more sane approach, but I think that's out of scope for core.

You would miss out on being able to detect coverage, but to address that we can do a better job of enforcing wrapping of all strings in future PRs and fixing them as we find them in existing code. As best practice for international users we should be doing this already, because the entire point of wrapping strings is to centralize their translation/phrasing via i18n!

**Comment by Griatch** [comment]
@strikaco The number of strings actually classified as "hard coded" under this scheme (that is, which is not sitting on Commands or Typeclasses or other entities meant to be changed by other means) is actually pretty limited. But it has still grown since this proposal was originally written some 3.5 years ago and my thinking has change since then too. In short - I agree with your objections here. 

> But...that's the entire point of the i18n module. 

No, i18n is, at least as far as I know, not intended to be used to *customize* strings, it's meant to translate strings *between languages*. But yes, compared to breaking variables out into a separate module using i18n is a saner way (assuming all our 'hard coded' strings actually were under i18n which they currently aren't). 

I'd been considering closing this for a good while; thanks for giving me the incentive to :)

**Comment by strikaco** [comment]
(Splitting hairs now) You're right, but what constitutes a language is arbitrary in this context. Chinese has a million dialects, plus standard and simplified variants. Portuguese has at least two. Even English breaks down into standard and simplified versions for British and Americans :)

My point is that regardless of its intended use case, it's flexible enough that it doesn't *have* to be only between languages; one could redefine English itself if they desired. If we're wrapping strings properly, users will always have the option of translating whatever they want into whatever language, dialect or phrasing they want.

**Comment by Griatch** [comment]
@strikaco I don't think there was ever any doubt that i18n *could* be used for this purpose. I will stand by the fact that it's not its intended usage and that relying on it in that way is not ideal. It's just likely better than any alternative (breaking out strings into separate modules or reworking every little bit of core to be pluggable just for the purpose of changing a string).
